File: src/brains/clustering/single_link.clj
(ns brains.clustering.single-link
  (use brains.clustering.util)
  (require [brains.clustering.dendrogram :as dgram]
           [clojure.contrib.seq :as seq]))

(defn build-clusters [d points]
  (loop [points points
         clusters nil]
    (if (empty? points)
      clusters
      (let [p (first points)
            others (rest points)
            parts (seq/separate #(> d (distance (:values p) (:values %)))
                                others)
            [close far] parts]
        (recur far (cons (cons p close) clusters))))))

(defn cluster [k points]
  (loop [clusters points
         distance 0
         d (dgram/new clusters)]
    (let [oldcount (count clusters)
          newclusters (build-clusters distance points)
          newcount (count newclusters)
          newdgram (if (not= oldcount newcount)
                     (dgram/add d distance newclusters)
                     d)]
      (if (= newcount k)
        newdgram
        (recur newclusters (+ distance 1) newdgram)))))

File: scala/src/main/scala/brains/clustering/SingleLink.scala
package brains.clustering

import scala.annotation.tailrec

object SingleLink extends Algorithm[brains.NumericDataPoint] with Driver {
  
  @tailrec 
  private 
  def buildClusters(distance:Double, points:Cluster, clusters:Clusters):Clusters = {
      if (points.isEmpty) clusters
      else {
        val p :: others = points.toList
        val (close, far) = others.partition(p.distance(_) < distance)
        buildClusters(distance, far, clusters :+ (close :+ p))
      }
  }

  def cluster(k:Int, points:Cluster):Clusters = {
    @tailrec
    def doCluster(clusters:Clusters, distance:Double):Clusters = {
      val newClusters = buildClusters(distance, points, Seq())
      if (newClusters.size <= k) newClusters
      else doCluster(newClusters, distance + 1)
    }

    val clusters = points map { Seq(_) }
    doCluster(clusters, 0)
  }

}

